client/windows/CHeroWindow.cpp

```diff
@@ -43,19 +43,21 @@
 const TBonusListPtr CHeroWithMaybePickedArtifact::getAllBonuses(const CSelector & selector, const CSelector & limit, const CBonusSystemNode * root, const std::string & cachingStr) const
 {
 	TBonusListPtr out(new BonusList());
-	TBonusListPtr heroBonuses = hero->getAllBonuses(selector, limit, hero, cachingStr);
+
+	// Fix bug: we are ignoring the cachingStr here, since when pick up and put back the artifact we want to refresh the cache
+	const TBonusListPtr heroBonuses = hero->getAllBonuses(selector, limit, hero);

```

Comment:
Revert back to using cachingStr.

Reply:
This line is actually fixing the issue, if we use cachingStr again, the result will be wrong, and the later codes cannot make it correct. So removing the cachingStr is necessary.

Reply:
Hi Alex, 
I believe this is the most straight forward (being simple, with minimum change, without any confusion) way to fix this issue, although it looked a little ugly, and lose efficiency on the cache.
Since the designer of the "cache" system didn't consider this "pick up" scenario, it might take much more effort to fix that with a neat solution, for such a small issue. And I suppose this (Pick up and put the art back) should be a corner case that makes it very special in the hero artifact system, so there should not be more cases like this.
Thanks for the consideration.

Reply:
I will look again on this issue (maybe today) if I will not find better solution I`ll merge this PR as is.